src/librustc_mir/borrow_check/nll/constraint_set.rs

```diff
@@ -0,0 +1,114 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use rustc::mir::Location;
+use rustc::ty::RegionVid;
+use rustc_data_structures::fx::FxHashSet;
+use rustc_data_structures::indexed_vec::{Idx, IndexVec};
+
+use std::fmt;
+use syntax_pos::Span;
+
+#[derive(Clone, Default)]
+crate struct ConstraintSet {
+    constraints: IndexVec<ConstraintIndex, OutlivesConstraint>,
+    seen_constraints: FxHashSet<(RegionVid, RegionVid)>,
+}
+
+impl ConstraintSet {
+    pub fn push(&mut self, constraint: OutlivesConstraint) {
+        debug!(
+            "add_outlives({:?}: {:?} @ {:?}",
+            constraint.sup, constraint.sub, constraint.point
+        );
+        if constraint.sup == constraint.sub {
+            // 'a: 'a is pretty uninteresting
+            return;
+        }
+        if self.seen_constraints.insert(constraint.dedup_key()) {
+            self.constraints.push(constraint);
+        }
+    }
+
+    pub fn inner(&self) -> &IndexVec<ConstraintIndex, OutlivesConstraint> {
```

Comment:
Nit: Instead of `Inner`, we could implement `IndexVec` for `ConstraintSet`. Something like this;

```rust
impl Index<ConstraintIndex> for ConstraintSet {
    type Output = OutlivesConstriant;

    fn index(&self, index: ConstraintIndex) -> &OutlivesConstraint {
        &self.constraints[index]
    }
}
```

then you can just do `constraint_set[index]` instead of `constraint_set.inner()[index]`